### ISSUE TITLE
Switch to little endian for serialized formats

### ISSUE THREAD

**Issue by Flamefire** [Return-To-The-Roots/s25client#1086]
Currently virtually all machines are little endian. Having our formats store values in big endian means a lot of conversion going on.

I propose to switch all serialized formats (replays, savegames, network messages) to little endian which will probably improve loading times and (slightly) run time.

Downside: Old files cannot be loaded, old and new clients/servers cannot communicate.

Proposed solution: Implement code to check for correct encoding/protocol version in lobby server first, then in clients/game server (simply DC in case of error). Then switch right before the next stable release (which should happen soon!)

@Flow86 Can you recheck all open issues and assign/remove to next milestone and release once those are reached? (PR first to double-check with other devs)

**Comment by Flow86** [comment]
since we still need the big endian loaders for libsiedler2, I dont think it makes much sense completely swap everything. (or isn't it used anymore?)

I would say, leave it as it is at the moment.
We could think of switching to google protobuf or similar for the network protocols so we do not need to manually implement the (un)marshalling anymore.

I think there's more interesting stuff to do than endianess?

**Comment by Flamefire** [comment]
> since we still need the big endian loaders for libsiedler2

Those are file formats and endianess there is handled explicitely as the file format requires it. So this is independent. What would be changed is `libutil::Serializer`: It has a `Converter` which can simply be switched to use little endian on file/target. On big endian machines it would convert then

> We could think of switching to google protobuf or similar for the network protocols so we do not need to manually implement the (un)marshalling anymore.

We also have the replays and savegames which can't be done with any other serialization library :/ I don't really like protobuf either as it adds a lot of complexity. The approach of Boost.Serialization is good (having 1 templated serialize method which is used for both directions) but the library itself is to heavy for my liking, but we could use that method.

Sure. This is just a low hanging fruit that's always been at the back of my head

**Comment by Flow86** [comment]
ah yes, boost::serialization is quiet interesting (but heavy) that could at least remove the custom made stuff.

> What would be changed is libutil::Serializer: It has a Converter which can simply be switched to use little endian on file/target. On big endian machines it would convert then

thats okay for me

**Comment by Flamefire** [comment]
> thats okay for me

Does that also include the network packets (for which the Serializer is also used) and which requires similar changes on the lobby server?

**Comment by Flow86** [comment]
yes you can do that if needed, if its also a "low hanging fruit"?

**Comment by Flamefire** [comment]
I just checked the server code and must say: Not so low.

The code base used is completely different to liblobby on Github. I was thinking about having the server detect, what protocol version is used and use little or big endian serializer depending on that. This would require to specify the serializer independently to the message but a message (strangely) is a serializer (design mistake)

Future-proof idea:
1. Unify the interfaces/repositories, so the lobby-server uses the same github repo as rttr (I already fixed that design bug there: the messages take a serializer as an argument)
2. Decide on a pattern to use for serialization, deserialization and dispatch
3. Maybe (depending on 2) provide different serializers

On 2: Currently multiple ways are used. 
1. `message_lobby_processor::process` dispatches to a virtual function using a switch over IDs
2. server lobby uses a factory to create messages where each message "registers itself"
3. client lobby uses a switch-case for creating
4. client lobby uses double-dispatch for dispatching (`run` method)

I think we can combine the switch-case with double dispatch into a single method:

```
template<typename F>
auto dispatcher(enum id, F&& func){
  switch(id){
    case FOO: return func((Foo*) nullptr); break;
   ...
   }
}

std::unique_ptr<Msg> create(id, data){
  return dispatcher(id, [&data](auto* null){ return std::make_unique<decltype(*null)>(data); });
}

bool run(msg, interface){
  return dispatcher(msg.id, [&interface](auto* null){ return interface.run(static_cast<decltype(null)>(data)); });
}
```

The only mapping of id to type would then be in `dispatcher`. Needs more work/though of course...

The idea would then be that we can have 2 implementations of `interface` (same as used for handling a message) for deserialization (and serialization) where one would use a little-endian serializer and the other a big-endian serializer.

Still all quite complicated if we want to support multiple serializer types as then the serialization routines must accept a template and we must put in the right type of the serializer through all the dispatching...

Obvious alternative: Prepare a PR for the lobbyserver which uses the new protocol and drops all old protocol client and deploy with the nightly of the client which uses the new protocol.

Related: https://codereview.stackexchange.com/questions/191257/packet-factory-design-for-networking-application
